### PR TITLE
New 'marking' feature that allows user to permanently highlight a data element in chart

### PR DESCRIPTION
**Pull request by joshhjacobson**
This is my first time contributing to an open source project on GitHub, so I hope I've done everything correctly. I've only implemented the feature in the parcoords.standalone.js file since I wasn't sure how to properly work it in universally. I've included an example in the demo folder as well. Please let me know if there's anything else I should do. Thanks!

### CODE REVIEW

**Comment by BigFatDog** [comment]
Thank you for your efforts for this pull request. I think this feature is useful.

I've tried the `marking.html` and got your idea. It would take me some time to go through your code.  

This is a good start, I'll come back after reading your code carefully.

**Comment by joshhjacobson** [comment]
Sounds great, thanks for taking a look! I'm happy to elaborate if necessary.

File: dist/parcoords.standalone.js [hunk 1]
```diff
@@ -8577,6 +8577,10 @@
         ctx.brushed.scale(devicePixelRatio, devicePixelRatio);
         ctx.highlight.lineWidth = 3;
         ctx.highlight.scale(devicePixelRatio, devicePixelRatio);
+        ctx.marked.lineWidth = 3;
+        ctx.marked.shadowColor = "#ffffff";
+        ctx.marked.shadowBlur = 10;
+        ctx.marked.scale(devicePixelRatio, devicePixelRatio);
```
Comment (BigFatDog) [inline dist/parcoords.standalone.js]: Hmm... Would you please make `shadowColor`, `shadowBlur` and marker's `lineWidth` in configurations?

I'll make highlight's `lineWidth` read from configuration later.

File: dist/parcoords.standalone.js [hunk 2]
```diff
@@ -9355,6 +9403,34 @@
       };
     };
 
+    // mark an array of data
+    var mark = function mark(config, pc, canvas, events, ctx, position) {
+      return function () {
+        var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
```
Comment (BigFatDog) [inline dist/parcoords.standalone.js]: [ES6's default parameter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters) can simplify this:
```
const highlight = (config, pc, canvas, events, ctx, position) =>
  function(data = null) {
    if (data === null) {
      return config.highlighted;
    }
```

**Comment by BigFatDog** [comment]
I have gone through your code. I want to merge this feature after it is written in source code other than `parcoords.standalone.js`.

I can convert your code, however, I think you are the right person to do this. If you do it, this feature is all your contribution as a whole. (That feels great as an open source contributor. 😃 )

Which approach do you like?

**Comment by joshhjacobson** [comment]
Thanks for reviewing this. I'd be happy to make the changes you mentioned and include additional source code. What all do you need? Based on structure of the repo, I'm guessing you'd like the feature to be implemented in `parcoords.js`, `parcoords.standalone.js`, and its own file `mark.js`.

File: dist/parcoords.standalone.js [hunk 3]
```diff
@@ -8766,7 +8770,7 @@
     var axisDots = function axisDots(config, pc, position) {
       return function (_r) {
         var r = _r || 0.1;
-        var ctx = pc.ctx.marks;
+        var ctx = pc.ctx.dots;
```
Comment (BigFatDog) [inline dist/parcoords.standalone.js]: corresponding line of source code:
https://github.com/BigFatDog/parcoords-es/blob/5959a719ec0447ce6864712ebfabe881ac65b85c/src/api/axisDots.js#L7

File: dist/parcoords.standalone.js [hunk 4]
```diff
@@ -8991,6 +8995,7 @@
           delete dragging[d];
           select(this).transition().attr('transform', 'translate(' + xscale(d) + ')');
```
Comment (BigFatDog) [inline dist/parcoords.standalone.js]: Corresponding line: https://github.com/BigFatDog/parcoords-es/blob/a035099c146c11998db13ab7b733f0f3996c5c59/src/api/reorderable.js#L33

**Comment by BigFatDog** [comment]
`parcoords.js` and `parcoords.standalone.js` are complied result of code under `src` folder. Therefore, all your changes (excluding ones to demo folder) should be applied to code that is in `src` folder.

I've linked source code to two of your changes. I will be glad to link all of your changes to sources if you need. It's up to you.

**Comment by joshhjacobson** [comment]
Ok, I think I can locate where most of the changes should be implemented and will do so over the next week. I'll reach out if I have any questions or issues with this. Should I submit a separate pull request with the new changes or just commit them to this one?

**Comment by BigFatDog** [comment]
Both of the two approaches are good to me... Maybe a separate pull request would be better? (because many code changes can be expected). What do you think?

**Comment by BigFatDog** [comment]
See you then.

File: dist/parcoords.standalone.js (hunk 1 above)
Comment (joshhjacobson) [reply dist/parcoords.standalone.js]: I've got the updated code almost finished. I'm not sure what you mean by making these in configurations, however. Could you sketch up the `highlight.linewidth` example in a reply so I can replicate that with marked?

Comment (BigFatDog) [reply dist/parcoords.standalone.js]: I mean end uses can customize `lineWidth`, `shadowColor` and `shadowBlur`.

Steps are:
1. Add definitions to [DefaultConfig](https://github.com/BigFatDog/parcoords-es/blob/develop/src/state/defaultConfig.js):
```
const DefaultConfig = {
...
  color: '#069',
  markerLineWidth: 3, 
  markerShadowColor: '#ffffff',
  markerShadowBlur: 10
...
}
```
Naming depends on your design: 
* If you plan to let marker feature share the global `lineWidth` or `shadowColor` with other features (normal line width, highlighted line width, etc.), please keep your original design, I'll update other features accordingly.
* If you'd like to reserve these config for marker feature only, `markerShadowBlur` should be a better choice.

2. Refer config. Here is an example: [usage of `dimensionTitleRotation`](https://github.com/BigFatDog/parcoords-es/blob/f41c1b1774b45718ff81890eb4fed8f010a365a8/src/api/updateAxes.js#L51)

3. End users customize:
```
pc()
.markerLineWidth(6)
.markerShadowColor('#ffffff')
.markerShadowBlur(10)
```

Comment (joshhjacobson) [reply dist/parcoords.standalone.js]: Ok great! I think I've got it all setup. I'm going to try to build and test the code and hopefully I'll have the pull request ready by the end of the day.

**Comment by joshhjacobson** [comment]
I've got all the code implemented in the src directory and was able to get a successful build with npm. Would you like me to include the parcoords.js and parcoords.esm.js files in the pull request, or just the code in src?

**Comment by BigFatDog** [comment]
Just the src is good enough.

**Pull request closed without merging**